### ISSUE TITLE
Non-blocking API in ConnectorPageSource

### ISSUE THREAD

**Issue by arhimondr** [prestodb/presto#5500]
It makes possible to use asynchronous data fetch in connectors. 

By analogy to the `ConnectorPageSink#appendPage` method, `ConnectorPageSource#getNextPage` should return `CompletableFuture<Page>` instead of `Page`. All of the connectors initially should be migrated to return `CompletableFuture.completedFuture(page)`. Supposed that wrapping the result into `completedFuture` won't do any harm to performance.

 If there are any performance related doubts, the `ListenableFuture<?> isBlocked()` method can be add to `ConnectorPageSource` interface with the default implementation of `return UNBLOCKED`.

The first approach is more clear and intuitive. Also it explicitly suggest to use unblocked algorithms for data fetching. The second approach won't introduce any harm to the performance for sure, and no connectors migration will be required.

@martint @dain What do you think? What approach is better to choose?

**Comment by arhimondr** [comment]
I choose second approach over the first one because of the following reasons:

1. It is backward compatible
2. It has no performance impact
3. With high probability non-blocking asynchronous  elements in connectors won't return `Future<Page>`, but some other `Future` that must the transformed before returning it to Presto. When the transformation of `Future` is trivial, it is not so for `ListenableFuture` that must call the underlying callbacks. It is not always clear what exactly thread is going to perform required transformation. And we must make sure that the Presto worker threads performs such operations (deserializing/encoding/etc.).

Actual implementation is here: https://github.com/prestodb/presto/pull/6343

**Comment by ilfrin** [comment]
#6343 was merged, closing.
